### PR TITLE
Add flagging utilities

### PR DESCRIPTION
**Pull request by abrookins**
### Summary
Add a feature flagging system so that developers can integrate in-progress work behind a flag.

This PR adds the core flagging utilities, tests, and docs. Future PRs will include:
- A CLI command
- A REST API to get flags (to be used by the UI)

Parent issue: #6244 

### Steps Taken to QA Changes
1. Set `PREFECT_FEATURE_FLAGGING_ENABLED=true` in my shell
2. Ran `ipython`
3. Created a feature flag
4. Checked if the feature flag existed
5. Enabled the flag
6. Checked if the flag was disabled

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- No tests or issue needed
- [ ] A short code fix
	- Please reference the related issue by including "closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a bug report issue 
	- Please include tests. One-line fixes without tests will not be accepted unless it's related to the documentation only.
- [x] A new feature implementation
	- Please reference the related issue by including "closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a feature enhancement issue 
	- Please include tests
    - Please make sure that your QA steps are both thorough and easy to reproduce by somebody with limited knowledge of the feature that you are submitting

**Happy engineering!**

### CODE REVIEW

File: docs/contributing/feature_flags.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+# Feature Flags
+
+This section describes how Prefect uses feature flagging during
+development.
+
+## What are feature flags?
+
+_Feature flagging_ is the process of introducing _flags_ that allow
+developers to turn a feature of an application on or off at runtime --
```
Comment (tpdorsey) [inline docs/contributing/feature_flags.md:16]:
```suggestion
developers to turn a feature of an application on or off at runtime &mdash;
```

File: docs/contributing/feature_flags.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+from flipper import Condition
+
+
+MY_FLAG = "my-flag"
+IS_ADMIN = Condition(is_admin=True)
+
+create_if_missing(MY_FLAG, conditions=[is_admin])
+```
+
+**NOTE**: You won't need to refer to the condition object later. This example
```
Comment (tpdorsey) [inline docs/contributing/feature_flags.md:98]:
```suggestion
!!! note "Referencing the condition object"
    You won't need to refer to the condition object later. This example
```

File: docs/contributing/feature_flags.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+includes it as a separate variable for ease of reading.
```
Comment (tpdorsey) [inline docs/contributing/feature_flags.md:99]:
```suggestion
    includes it as a separate variable for ease of reading.
```

File: docs/contributing/feature_flags.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+# `flag_is_enabled` is that this time, we pass `is_admin=True`.
+if feature_flags.flag_is_enabled(feature_flags.MY_FLAG, is_admin=True):
+    # your_special_logic()
+    ...
+else:
+    # your_normal_logic()
+    ...
+```
+
+**NOTE**: Passing conditions as input to `flag_is_enabled()` is not required
```
Comment (tpdorsey) [inline docs/contributing/feature_flags.md:118]:
```suggestion
Passing conditions as input to `flag_is_enabled()` is not required
```

File: docs/contributing/feature_flags.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+when you check if a flag that has conditions is enabled. If you don't specify
+any conditions to `flag_is_enabled()`, the return value is the current state
+of the flag, e.g. True if the flag is enabled.
```
Comment (tpdorsey) [inline docs/contributing/feature_flags.md:121]:
```suggestion
of the flag, e.g. `True` if the flag is enabled.
```

File: docs/contributing/feature_flags.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+## Turning a flag on and off for specific "buckets"
```
Comment (tpdorsey) [inline docs/contributing/feature_flags.md:123]:
```suggestion
## Toggle a flag for specific "buckets"
```

File: src/prefect/utilities/feature_flags.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
+"""
+This module contains code related to feature flagging.
+
+For more guidance, read docs/feature_flags.md.
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:4]:
```suggestion
For more guidance, read the [Feature Flags](/contributing/feature_flags/) documentation.
```

File: src/prefect/utilities/feature_flags.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    bucketer: Optional[AbstractBucketer] = None,
+    conditions: Optional[Iterable[Condition]] = None,
+    client: FeatureFlagClient = None,
+) -> Optional[FeatureFlag]:
+    """
+    Create a feature flag if a flag matching the given name does not
+    already exist.
+
+    Args:
+        flag_name: the name of the feature flag
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:42]:
```suggestion
        flag_name: The name of the feature flag.
```

Comment (tpdorsey) [reply src/prefect/utilities/feature_flags.py:42]: Sentence caps and punctuation on the descriptions, please.

File: src/prefect/utilities/feature_flags.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        is_enabled: the initial enabled/disabled state of the flag if
+                    this function creates it
+        client_data: arbitrary data that we should store with the flag
+        bucketer: an optional bucketer from the flipper.bucketing module, e.g.
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:46]:
```suggestion
        bucketer: an optional bucketer from the `flipper.bucketing` module, e.g.
```

File: src/prefect/utilities/feature_flags.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+                  PercentageBucketer, to use when determining if the flag
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:47]:
```suggestion
                  `PercentageBucketer`, to use when determining if the flag
```

File: src/prefect/utilities/feature_flags.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+                  is enabled
+        conditions: an optional iterable of Conditions against which we will
+                    check input data to determine if a flag is enabled
+        client: The FeatureFlagClient instance to use. Defaults to a client
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:51]:
```suggestion
        client: The `FeatureFlagClient` instance to use. Defaults to a client
```

File: src/prefect/utilities/feature_flags.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+    Args:
+        flag_name: the name of the feature flag
+        is_enabled: the initial enabled/disabled state of the flag if
+                    this function creates it
+        client_data: arbitrary data that we should store with the flag
+        bucketer: an optional bucketer from the flipper.bucketing module, e.g.
+                  PercentageBucketer, to use when determining if the flag
+                  is enabled
+        conditions: an optional iterable of Conditions against which we will
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:49]: Does "Conditions" reference something that is a proper name? If not can we lowercase? I'm not familiar with the usage.

File: src/prefect/utilities/feature_flags.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+    flag_name: str,
+    default=False,
+    client: FeatureFlagClient = None,
+    **conditions: Optional[Any]
+):
+    """
+    Check if a feature flag is enabled.
+
+    This function always returns False if the setting
+    PREFECT_CLOUD_ENABLE_FEATURE_FLAGGING is false.
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:91]:
```suggestion
    `PREFECT_CLOUD_ENABLE_FEATURE_FLAGGING` is false.
```

File: src/prefect/utilities/feature_flags.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+
+    NOTE: If `flag_is_enabled()` is called for a feature that has conditions,
+    but the caller does not give any conditions, the current state of the flag
+    is returned.
+
+    Args:
+        flag_name: the name of the feature flag
+        default: the default return value to use if no feature flag with
+                 the given name exists. Defaults to False.
+        client: The FeatureFlagClient instance to use. Defaults to a client
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:101]:
```suggestion
        client: The `FeatureFlagClient` instance to use. Defaults to a client
```

File: src/prefect/utilities/feature_flags.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+                configured to look at an in-memory feature store.
+        conditions: keyword arguments, e.g. is_admin=True, to check
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:103]:
```suggestion
        conditions: keyword arguments, e.g. `is_admin=True`, to check
```

File: src/prefect/utilities/feature_flags.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+
+def list_feature_flags(
+    batch_size: int = 10, client: FeatureFlagClient = None
+) -> List[FeatureFlag]:
+    """
+    List all feature flags.
+
+    This function always returns an empty list if the setting
+    PREFECT_CLOUD_ENABLE_FEATURE_FLAGGING is false.
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:125]:
```suggestion
    `PREFECT_CLOUD_ENABLE_FEATURE_FLAGGING` is false.
```

File: src/prefect/utilities/feature_flags.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+    Args:
+        batch_size: batch size of flags to retrieve at a time
+        client: The FeatureFlagClient instance to use. Defaults to a client
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:129]:
```suggestion
        client: The `FeatureFlagClient` instance to use. Defaults to a client
```

**Review by tpdorsey — COMMENTED** [review]
Looks good. Some minor edits and suggestions.

For these new pages to show up in the docs, a few extra steps:

1. Create a docs/api-ref/prefect/utilities/feature_flags.md file. See neighboring files for the format of the contents. It points to the source import.
2. Add docs/contributing/feature_flags.md and docs/api-ref/prefect/utilities/feature_flags.md to the `nav` section in mkdocs.yml.

Ping me if you need any assistance with this.

**Comment by madkinsz** [comment]
@abrookins do you have time to finish this one up?

**Comment by abrookins** [comment]
@madkinsz Realistically, no, but I'm going to give it a college try.

File: src/prefect/utilities/feature_flags.py (hunk 12 above)
Comment (abrookins) [reply src/prefect/utilities/feature_flags.py:52]: `Condition` is a class name, so it should probably get the backtick treatment.

File: src/prefect/utilities/feature_flags.py [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+    Args:
+        flag_name: The name of the feature flag.
+        is_enabled: the initial enabled/disabled state of the flag if
+                    this function creates it
+        client_data: arbitrary data that we should store with the flag
+        bucketer: an optional bucketer from the `flipper.bucketing` module, e.g.
+                  `PercentageBucketer`, to use when determining if the flag
+                  is enabled
+        conditions: an optional iterable of Conditions against which we will
```
Comment (abrookins) [inline src/prefect/utilities/feature_flags.py:52]:
```suggestion
        conditions: an optional iterable of `Condition` instances against which we will
```

File: src/prefect/utilities/feature_flags.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+
+    Args:
+        flag_name: The name of the feature flag.
+        is_enabled: the initial enabled/disabled state of the flag if
+                    this function creates it
+        client_data: arbitrary data that we should store with the flag
+        bucketer: an optional "bucketter" from the `flipper` module, e.g.
+                  `PercentageBucketer`, to use when determining if the
+                  flag is enabled
+        conditions: an optional iterable of `Condition` instances from the
```
Comment (abrookins) [inline src/prefect/utilities/feature_flags.py:52]: @tpdorsey Want to check out this docstring again? There are some difficult-communicate subtleties here that the type annotations explain better (e.g., `AbstractBucketer` is an interface, but `Condition` is an instance).

File: src/prefect/utilities/feature_flags.py [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+                    `flipper` module against which we will check input data
+                    to determine if a flag is enabled
+        client: The `FeatureFlagClient` instance to use. Defaults to a client
+                configured with an in-memory feature store.
+
+    Returns:
+        FeatureFlag or None: Returns a created or existing FeatureFlag, or None
```
Comment (abrookins) [inline src/prefect/utilities/feature_flags.py:59]:
```suggestion
        `FeatureFlag` or None: Returns a created or existing `FeatureFlag`, or None
```

File: src/prefect/utilities/feature_flags.py [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+        return flags
+
+
+def get_default_feature_flagger() -> FeatureFlagger:
+    """
+    Return the default feature flagger.
+
+    The default feature flagger uses an in-memory flag store.
+
+    Once the feature flag
```
Comment (tpdorsey) [inline src/prefect/utilities/feature_flags.py:197]: Incomplete sentence

**Pull request closed without merging**